packages/internals/support/src/entities/component.js

```diff
@@ -1,101 +1,120 @@
-const assert = require('check-types').assert;
-const check = require('check-more-types');
-const {omit} = require('lodash');
-const Collection = require('../collections/collection');
+const {normalizeName, uniqueName} = require('@frctl/utils');
+const Validator = require('../validator');
+const schema = require('../../schema');
+const VariantCollection = require('../collections/variant-collection');
 const FileCollection = require('../collections/file-collection');
 const Entity = require('./entity');
-const File = require('./file');
-
-const isFile = value => value instanceof File;
-check.mixin(isFile, 'file');
-const isCollection = value => value instanceof Collection;
-check.mixin(isCollection, 'collection');
-const isFileCollection = value => value instanceof FileCollection;
-check.mixin(isFileCollection, 'fileCollection');
-
-const componentSchema = {
-  path: check.maybe.unemptyString,
-  relative: check.maybe.unemptyString,
-  src: check.file,
-  name: check.maybe.unemptyString,
-  config: check.maybe.object,
-  variants: check.maybe.collection,
-  files: check.maybe.fileCollection
-};
+const Variant = require('./variant');
 
 const _src = new WeakMap();
 const _files = new WeakMap();
 const _variants = new WeakMap();
 
-const privateProps = ['src', 'files', 'variants'];
-
 class Component extends Entity {
-  constructor(props = {}) {
-    assert(check.schema(componentSchema, props), `Component.constructor: The properties provided do not match the schema of a component [properties-invalid]`, TypeError);
+  constructor(props) {
+    if (Component.isComponent(props)) {
+      return props;
+    }
+    Component.validate(props);
 
-    props.name = props.name || props.src.stem;
-    props.config = props.config || {};
+    const config = Object.assign({
+      name: normalizeName(props.src.stem)
+    }, props.config);
 
-    super(omit(props, privateProps));
+    super(config);
 
-    _src.set(this, props.src);
-    _files.set(this, props.files || new FileCollection());
-    _variants.set(this, props.variants || new Collection());
+    this._setSrc(props.src);
+    this._setFiles(props.files);
+    this._buildVariants(config.variants);
+  }
 
-    // TODO: Revisit if memory impacted
-    Object.defineProperty(this, 'path', {
-      enumerable: true,
-      get: pathGet
-    });
+  getSrc() {
+    return _src.get(this).clone();
+  }
 
-    Object.defineProperty(this, 'relative', {
-      enumerable: true,
-      get: relativeGet
-    });
+  getFiles() {
+    return _files.get(this).clone();
+  }
 
-    Object.defineProperty(this, 'files', {
-      enumerable: true,
-      get: filesGet,
-      set: filesSet
-    });
+  addFile(file) {
+    _files.set(this, _files.get(this).push(file));
+  }
+
+  getVariants() {
+    return _variants.get(this);
+  }
+
+  getVariant(name) {
+    return _variants.get(this).find(name);
+  }
+
+  addVariant(variant) {
```

Comment:
I think that people should be able to supply a plain object here and have the component create the Variant instance from it (so that they don't need to add the component name property, generate a name etc). Actually I'd say that maybe that functionality could live on the VariantCollection itself (i.e. pushing/concating an object would auto-instantiate a Variant) but that could open up a whole can of worms with regards to the other collections....! Maybe just have the component create the variant and then add it to the VariantCollection for now?

Reply:
Yes, I was wondering about that. Sounds good. 👍